packages/grid/_modules_/grid/models/gridSlotsComponentsProps.ts

```diff
@@ -1,10 +1,13 @@
+import type { GridColumnsHeaderProps } from '../components/columnHeaders/GridColumnHeaders';
+
 /**
  * Overrideable components props dynamically passed to the component at rendering.
  */
 export interface GridSlotsComponentsProps {
   checkbox?: any;
   columnMenu?: any;
   columnsPanel?: any;
+  columnsHeader?: GridColumnsHeaderProps;
```

Comment:
Why did you add a new prop? There's no respective slot for it.

Reply:
It's related to 

>  The components should never import selectors from optional hook like useGridColumnReorder, useGridColumnResize and useGridInfiniteLoader. Instead, the feature hook should inject the props in props.componentsProps?.XXX

Reply:
It's also related to https://github.com/mui-org/material-ui-x/issues/2293#issuecomment-914266340.

The purpose of `componentsProps` was to allow to change the props of inner components without having to recreate the entire grid. However, the `GridColumnsHeader` has no prop for the user to change. I agree to add new options to `GridSlotsComponentsProps` for those components that could be relevant to allow users to change their props, but not for internal components that only we need to change.

To provide an alternative, we can use events to get the current column being dragged/resized. In these two cases, I was already advocating that these keys could be removed from the state and replaced with events. See https://github.com/mui-org/material-ui-x/issues/820#issuecomment-897906608 and https://github.com/mui-org/material-ui-x/pull/2337

For the column resizing, the selector is first called in `GridColumnHeadersItemCollection` but its value is only used in `GridColumnHeaderSeparator`. The tree is: 

```
GridColumnHeadersItemCollection -> GridColumnHeaderItem -> GridColumnHeaderSeparator
```

For column dragging, it doesn't change too much. The selector is called in `GridColumnsHeader`. Then, its value is used in this first component and passed to `GridColumnHeaderItem`. There's one component in the middle that doesn't need it. It's not a complex task to replace with events. The three is:

```
GridColumnsHeader -> GridColumnHeadersItemCollection -> GridColumnHeaderItem
```

What do you think about using events?

Reply:
I think its worth taking a look at the future pro-only features and see if the information we want to pass to the components can easily be implemented via events.
To avoid having to find another mechanism in a few weeks.

But on paper, events are probably a good approach.
As long as we accept that the event enum is on the `x-data-grid` version, even for paid-only events.
Which won't be a problem for bundle size.
And for separation of concern, it's not perfect, but if we wanted to really put all the code related to a paid-only feature inside the pro package, then we would have the same discussions for the localization and I don't think we want to split it.

I'll try an event-based implementation tomorrow to see how it looks.

Reply:
Another PR only removing the selector and listening to the event is better. There's no need to remove the key from the state right now.

Reply:
To merge this PR, I need to remove the selector from the component because the state is never set on the DataGrid, only on the DataGridPro now that its the hook which initialize its own state.

But I can do it on another PR, it just need to be merged before this one.

Reply:
See #2716